### PR TITLE
upgrade deseq2 and include tximport to use sailfish results directly

### PR DESCRIPTION
**Pull request by bgruening**
This one is called DESeq2-TNG I should have named it RNA-seq TNG ;)
I'm very excited about this PR and I'm seeking for help in testing. A little bit of background: until now we need to map the reads and use htseq-count to count reads in a region to feed this into DESeq2. 
On the other hand [Sailfish](http://www.cs.cmu.edu/~ckingsf/software/sailfish/downloads.html) is a "rapid Alignment-free Quantification of Isoform Abundance" tool ... but we can not use the output natively with DESeq2. This has changed now! 
@mikelove has developed a nice tool called [tximport](https://github.com/mikelove/tximportData) that can read sailfish and salmon outputs directly and does the DESEq2 magic on top of it.

This PR updated our DESEq2 wrapper to make use of this new functionality. The workflow should be much simpler and much faster, as it is alignment free. I have created a new DESEq2 conda package to ease testing. So please activate the "bgruening" conda channel and run planemo on it if you want to provide feedback!

@mikelove and me would love to have as much feedback as possible until the new bioc release in two weeks and the new DESeq2 release. After this there is a freeze for the next 6 month ... more or less.

Let us know what you think we both are pretty excited about this :)
Cheers,
Michael and Bjoern

### CODE REVIEW

File: tools/deseq2/deseq2.xml [hunk 1]
```diff
@@ -75,6 +93,29 @@
...
+        <conditional name="tximport">
+            <param name="tximport_selector" type="select" label="Choice of Input data">
+                <option value="count" selected="True">Count data (e.g. from htseq-count or feature-count)</option>
+                <option value="tximport">TPM values (e.g. from sailfish or salmon)</option>
+            </param>
+            <when value="tximport">
+                <conditional name="mapping_format">
+                    <param name="mapping_format_selector" type="select" label="Gene mapping format">
+                        <option value="gtf" selected="True">GTF</option>
+                        <option value="tabular">Transkript-ID and Gene-ID mapping file</option>
```
Comment (mvdbeek) [inline tools/deseq2/deseq2.xml]: s/Transkript/Transcript/

File: tools/deseq2/deseq2.xml [hunk 2]
```diff
@@ -75,6 +93,29 @@
...
+                    </param>
+                    <when value="gtf">
+                        <param name="gtf_file" type="data" format="gtf" optional="True" label="GTF file with Transkript - Gene mapping"/>
+                    </when>
+                    <when value="tabular">
+                        <param name="tabular_file" type="data" format="tabular" optional="True" label="Tabular file with Transkript - Gene mapping"/>
```
Comment (mvdbeek) [inline tools/deseq2/deseq2.xml]: s/Transkript/Transcript/

**Comment by mvdbeek** [comment]
Hmm, so if I understand well tximport is producing a gene-level count-table from kallisto/sailfish/ etc ... ?
Wouldn't it be more useful to have a separate tximport wrapper?
Then we could also use edgeR/limma/voom with rsem/sailfish/kallisto

**Comment by mvdbeek** [comment]
This is definitely cool, I'll give it a try.
Is there any of these "non-aligners" available in galaxy?

**Comment by bgruening** [comment]
@mvdbeek sailfish is available. Thanks for the review.

**Comment by bgruening** [comment]
Btw. the conda package is in my repo for now because it's all in the bioc devel channel. I will submit this to bioconda as soon as it is officially released.
A tool_dependency is still missing ... if anyone ... :)

ping @davebx @nekrut @ pvanheus

**Comment by mvdbeek** [comment]
tools/deseq2/test-data/deseq2_out.tab has changed significantly ... could you perhaps replace the full output file with the top 100 list and use `compare="contains"` ?
Also, the results of deseq2 often changes minimally in the last digits after ".".
When I'm testing our (deprecated) deseq wrappers, I tend to use `compare=re_match` on a shortened gene count list, where I replace 0.12345123423423432 with 0\.123.+ .
I guess this would give more meaning to the actual test-data, if from release to release the data is roughly the same.

**Comment by bgruening** [comment]
@mvdbeek I can do this hopefully in the next days. If you are faster please feel free to PR to this branch.

**Comment by mvdbeek** [comment]
@bgruening no problem, I already got a dirty script to do it ;)

**Comment by bgruening** [comment]
I'm updating the conda wrappers right now to fix the script and waiting for the USA guys to chime in :)

**Comment by mikelove** [comment]
@mvdbeek yes, tximport is designed to be modular and import Sailfish/Salmon/kallisto/RSEM and work with DESeq2/edgeR/limma or any other tool that takes a count matrix

@bgruening there is a 6 month freeze on the deseq2.R script which lives in the Bioconductor *release branch*. I wonder, would it make sense to put this deseq2.R script into the Galaxy repo as well, so that we can have continual development and bug fixes? I could fix things in devel and then you could pull from the devel branch as often as you like... just a thought.

BTW here is the tximport vignette: 

http://master.bioconductor.org/packages/3.3/bioc/vignettes/tximport/inst/doc/tximport.html

Regarding modularity, one subtlety is that there are two ways of passing the summarized data to the DE count tools: either pass a count matrix plus an offset matrix (compatible with DESeq2/edgeR) or pass a scaled count matrix (compatible with any count tool). Both approaches correct for changes in gene length across samples. I prefer the first approach a bit, but Charlotte Soneson the first author on the article on this topic](http://f1000research.com/articles/4-1521/v1) showed that both approaches work. If you wanted to make things easy and to have tximport work with any downstream count tool, you could use the tximport argument `countsFromAbundance="lengthScaledTPM"`. The output would just be a count matrix which could be imported by any DE tool.

**Comment by nekrut** [comment]
@davebx @martenson can we have on test soon please

**Comment by martenson** [comment]
@nekrut Test is off-limits afaik because of a new tool deployment process that only @natefoo knows

**Comment by bgruening** [comment]
This should not be merged. It only has conda dependencies and the dependencies are beta. They will only be ready with a new conda release. I had opened this one because I was seeking for comments and testers. It would be good if we can have this on test! But if it needs to be merged before I would wait until someone has created a tool_dependency for it. Test is not running conda resolver isn't it?

**Comment by bgruening** [comment]
Anyone know how I can reopen this PR?

**Comment by martenson** [comment]
@bgruening you cannot, you have to create a new one

**Pull request merged**